src/Feldspar/Core/Middleend/OptimizeUntyped.hs

```diff
@@ -1,81 +1,97 @@
+ , v1 == v3 = go env e3
 
 -- Same rule as previous rule but with Elements as backing write.
-go (In (App GetIx _ [arr, In (Literal (LInt _ _ n))]))
+go env (In (App GetIx _ [arr, In (Literal (LInt _ _ n))]))
  | In (App EMaterialize _ [In Literal{}, e@(In (App EPar _ _))]) <- arr
- , Just e3 <- grabWrite n e = go e3
+ , Just e3 <- grabWrite n e = go env e3
 
 -- Tuple selections, 1..15. Deliberately avoiding take 1 . drop k which will
 -- result in funny things with broken input.
-go (In (App Sel1 _  [In (App Tup _ (e:_))]))                             = go e
-go (In (App Sel2 _  [In (App Tup _ (_:e:_))]))                           = go e
-go (In (App Sel3 _  [In (App Tup _ (_:_:e:_))]))                         = go e
-go (In (App Sel4 _  [In (App Tup _ (_:_:_:e:_))]))                       = go e
-go (In (App Sel5 _  [In (App Tup _ (_:_:_:_:e:_))]))                     = go e
-go (In (App Sel6 _  [In (App Tup _ (_:_:_:_:_:e:_))]))                   = go e
-go (In (App Sel7 _  [In (App Tup _ (_:_:_:_:_:_:e:_))]))                 = go e
-go (In (App Sel8 _  [In (App Tup _ (_:_:_:_:_:_:_:e:_))]))               = go e
-go (In (App Sel9 _  [In (App Tup _ (_:_:_:_:_:_:_:_:e:_))]))             = go e
-go (In (App Sel10 _ [In (App Tup _ (_:_:_:_:_:_:_:_:_:e:_))]))           = go e
-go (In (App Sel11 _ [In (App Tup _ (_:_:_:_:_:_:_:_:_:_:e:_))]))         = go e
-go (In (App Sel12 _ [In (App Tup _ (_:_:_:_:_:_:_:_:_:_:_:e:_))]))       = go e
-go (In (App Sel13 _ [In (App Tup _ (_:_:_:_:_:_:_:_:_:_:_:_:e:_))]))     = go e
-go (In (App Sel14 _ [In (App Tup _ (_:_:_:_:_:_:_:_:_:_:_:_:_:e:_))]))   = go e
-go (In (App Sel15 _ [In (App Tup _ (_:_:_:_:_:_:_:_:_:_:_:_:_:_:e:_))])) = go e
+go env (In (App Sel1 _  [In (App Tup _ (e:_))]))
+  = go env e
+go env (In (App Sel2 _  [In (App Tup _ (_:e:_))]))
+  = go env e
+go env (In (App Sel3 _  [In (App Tup _ (_:_:e:_))]))
+  = go env e
+go env (In (App Sel4 _  [In (App Tup _ (_:_:_:e:_))]))
+  = go env e
+go env (In (App Sel5 _  [In (App Tup _ (_:_:_:_:e:_))]))
+  = go env e
+go env (In (App Sel6 _  [In (App Tup _ (_:_:_:_:_:e:_))]))
+  = go env e
+go env (In (App Sel7 _  [In (App Tup _ (_:_:_:_:_:_:e:_))]))
+  = go env e
+go env (In (App Sel8 _  [In (App Tup _ (_:_:_:_:_:_:_:e:_))]))
+  = go env e
+go env (In (App Sel9 _  [In (App Tup _ (_:_:_:_:_:_:_:_:e:_))]))
+  = go env e
+go env (In (App Sel10 _ [In (App Tup _ (_:_:_:_:_:_:_:_:_:e:_))]))
+  = go env e
+go env (In (App Sel11 _ [In (App Tup _ (_:_:_:_:_:_:_:_:_:_:e:_))]))
+  = go env e
+go env (In (App Sel12 _ [In (App Tup _ (_:_:_:_:_:_:_:_:_:_:_:e:_))]))
+  = go env e
+go env (In (App Sel13 _ [In (App Tup _ (_:_:_:_:_:_:_:_:_:_:_:_:e:_))]))
+  = go env e
+go env (In (App Sel14 _ [In (App Tup _ (_:_:_:_:_:_:_:_:_:_:_:_:_:e:_))]))
+  = go env e
+go env (In (App Sel15 _ [In (App Tup _ (_:_:_:_:_:_:_:_:_:_:_:_:_:_:e:_))]))
+  = go env e
 
 -- Fallthrough.
-go (In (App p t es)) = In (App p t $ map go es)
+go env (In (App p t es)) = In (App p t $ map (go env) es)
```

Comment:
The environment cannot communicate information between sub-expressions, but maybe you don't need that

Reply:
I have a few more changes planned for the optimizer but nothing I have needs that.